### PR TITLE
ECSW-2754: GitHub action for publishing provider node package and binary

### PR DESCRIPTION
**Pull request by nkoturovic**
## Description

GitHub action that publishes the `pulumi-stackpath` provider (binary and NodeJS SDK), which is being triggered by the push-tag event. The binary is published via GitHub release, and NodeJS SDK is uploaded to the official StackPath's public NPM repository.

**TODO:** 

* [x] Provide access to stackpath's public NPM credentials with safety considerations (public GitHub repo)
* ~[ ] Ensure token security~
    * ~[ ] Idea: Limit tagging to, and protect: `release/*` and `main/*`, plus protect `.github/workflows` folder, and restrict pushes to this folder by public users (on any branch). This way, only org users or code owners will be able to update `.github/workflow` and trigger the action by pushing tags to one of the protected `main`, `release/*` branches~. 

### Security concerns rationale:

Users outside of the organization can create pull requests only on their forks of the repo, and therefore they won't have access to secrets, and they will be able to create GitHub releases only on their forks also.   

Link to explanation: https://stackpath.slack.com/archives/C01VDTN0Q73/p1707325280032109

## References

* Jira ticket: [ECSW-2574](https://stackpath.atlassian.net/browse/ECSW-2574)
* https://github.com/stackpath/pulumi-stackpath/pull/3

[ECSW-2574]: https://stackpath.atlassian.net/browse/ECSW-2574?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by JDTX0 — DISMISSED** [review]
Just a few comments on some versions and a suggestion on checkout. nice work!

File: .github/workflows/build.yaml [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  push:
+    tags:
+      - 'v*'
+jobs:
+  publish_binary:
+     name: publish          
+     runs-on: ubuntu-latest 
+     steps:
+     - name: Checkout Repo
+       uses: actions/checkout@v2
```
[the remark this answers is not in the document]
Comment (nkoturovic) [reply .github/workflows/build.yaml]: Addressed in: 57269a3

File: .github/workflows/build.yaml [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+     - name: Unshallow clone for tags
+       run: git fetch --prune --unshallow --tags
```
[the remark this answers is not in the document]
Comment (nkoturovic) [reply .github/workflows/build.yaml]: Addressed in: 57269a3

File: .github/workflows/build.yaml [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+     - name: Install Go
+       uses: actions/setup-go@v2
```
[the remark this answers is not in the document]
Comment (nkoturovic) [reply .github/workflows/build.yaml]: Addressed in: ba816b9

File: .github/workflows/build.yaml [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+       with:
+         go-version: ${{matrix.goversion}}
+     - name: Install pulumictl
+       uses: jaxxstorm/action-install-gh-release@v1.5.0
+       with:
+         repo: pulumi/pulumictl
+     - name: Run GoReleaser
+       uses: goreleaser/goreleaser-action@v2
```
[the remark this answers is not in the document]
Comment (nkoturovic) [reply .github/workflows/build.yaml]: Addressed in: ba816b9

**Comment by nkoturovic** [comment]
Public contributions are only possible via forking the original repo, which eliminates security concerns.
Due to the description, the `go-lint` action can be added, but it should not be addressed here but as part of another ticket.

**Comment by nkoturovic** [comment]
@JDTX0 Thank you for the great comments!

**Comment by nkoturovic** [comment]
@JDTX0 The action started failing after I bumped versions:
https://github.com/stackpath/pulumi-stackpath/actions/runs/7844258707

I can either:
1. Revert to older versions
2. Update https://github.com/stackpath/actions to allow restricted workflows

**Review by binodluitel — APPROVED** [review]

**Comment by JDTX0** [comment]
@nkoturovic You can update this line in admin to allow the action:
https://github.com/stackpath/admin/blob/main/actions/index.ts#L40

Should be fine to update it to use `*` to allow any version so we don't have to mess with it in the future.

**Pull request merged**